### PR TITLE
file.d: add overflow checks

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -293,10 +293,17 @@ version (Posix) private void[] readImpl(const(char)[] name, const(FSChar)* namez
...
+        size = addu(size, actual, overflow);
+        if (overflow) assert(0);
+
         if (size >= upTo) break;
         if (size < result.length) continue;
-        immutable newAlloc = size + sizeIncrement;
+        immutable newAlloc = addu(size, sizeIncrement, overflow);
+        if (overflow) assert(0);
+
         result = GC.realloc(result.ptr, newAlloc, GC.BlkAttr.NO_SCAN)[0 .. newAlloc];
```
Comment (MartinNowak) [inline std/file.d]: Why?
You'll always crash before reaching that overflow. It's a size_t, so your address space is gone before it overflows.
That's too much ugly code for it's uselessness.

**Comment by MartinNowak** [comment]
See, this is the mess w/ all those crappy tiny PRs.
Now we having plenty of separate discussions #4712, #4713, #4714, #4715, #4716, #4717, #4718 about the same thing. This literally feels like sneaking in changes to me.

**Comment by MartinNowak** [comment]
Where does the sudden interest in size_t overflows come from? In particular when used in buffer resizing there seems to be no practical use, b/c as pointed out above, you'll always end up w/o address space long before the overflow.

**Comment by MartinNowak** [comment]
Before flooding phobos w/ even more of this ugly code, please write a small `SafeInt` helper in std/internal that does those checks.

**Comment by schveiguy** [comment]
> you'll always end up w/o address space long before the overflow.

Not necessarily. In many cases, you request a malloc for the result of a multiplication. If it overflows, the size actually granted could succeed and be quite small. But because malloc only tells you that you succeeded and here's your pointer, it's up to you to make that into a safe D array. In other words, your code thinks you requested space for 2 billion `int`, but instead you got space for 50 million. However, you just happily slice the pointer for 2 billion having a large unsafe access to all address space. In 64 bits, the overflow case is much more unlikely.

For the addition overflows, I agree the benefit is debatable. But if you are going to check for one case, you may as well check for all.

I do agree that a `SafeInt` type would look MUCH better than all these calls. It's easy to miss some operations too.

**Comment by JackStouffer** [comment]
> I do agree that a SafeInt type would look MUCH better than all these calls. It's easy to miss some operations too without such a nice wrapper type.

Then review https://github.com/dlang/phobos/pull/4613

File: std/file.d (hunk 1 above)
Comment (WalterBright) [inline std/file.d]: > You'll always crash before reaching that overflow.

Famous last words :-)

All assert's are guaranteed to never be tripped. Of course, we're often wrong about that, and they get tripped. Hence adding this here.

**Comment by JackStouffer** [comment]
@andralex I thought the current rule was "no std.experimental use in std".

**Comment by andralex** [comment]
@JackStouffer on the contrary, use of experimental facilities in phobos is good dogfooding.

**Comment by WalterBright** [comment]
> Simply replacing size_t size = 0; with auto size = checked!size_t(0); and leaving all other code unchanged takes care of it.

Turns out it's more than that, as the diffs show.

**Comment by WalterBright** [comment]
Hmm, now I get:
```
object.Error@src/rt/minfo.d(371): Cyclic dependency between module std.experimental.allocator and std.concurrency
std.experimental.allocator* ->
std.conv ->
std.datetime ->
std.datetime.timezone ->
std.concurrency* ->
std.stdio ->
std.file ->
std.experimental.checkedint ->
std.experimental.allocator.common ->
std.experimental.allocator*
```
checkedint should not be having cycles

**Comment by andralex** [comment]
Time to get rid of them `static shared this()`. https://github.com/dlang/phobos/pull/5421

**Comment by andralex** [comment]
https://github.com/dlang/phobos/pull/5423, too

File: std/file.d [hunk 2]
```diff
@@ -356,24 +356,26 @@ version (Posix) private void[] readImpl(const(char)[] name, const(FSChar)* namez
         : minInitialAlloc));
     void[] result = uninitializedArray!(ubyte[])(initialAlloc);
     scope(failure) GC.free(result.ptr);
-    size_t size = 0;
+
+    import std.experimental.checkedint;
+    auto size = checked!Abort(cast(size_t)0);
```
Comment (andralex) [inline std/file.d]: Abort is the default, please use `checked(size_t(0))`

File: std/file.d [hunk 3]
```diff
@@ -2644,7 +2646,11 @@ version(Windows) string getcwd()
         3. the buffer (lpBuffer) is not large enough: the required size of
     the buffer, in characters, including the null-terminating character.
     */
-    wchar[4096] buffW = void; //enough for most common case
+    version (unittest)
+        enum BUF_SIZE = 10;     // trigger reallocation code
+    else
+        enum BUF_SIZE = 4096;   // enough for most common case
```
Comment (andralex) [inline std/file.d]: neat idea

File: std/file.d [hunk 4]
```diff
@@ -2654,10 +2660,12 @@ version(Windows) string getcwd()
     }
     else //staticBuff isn't enough
     {
-        auto ptr = cast(wchar*) malloc(wchar.sizeof * n);
+        import std.experimental.checkedint;
+        auto cn = checked!Abort(n);
```
Comment (andralex) [inline std/file.d]: `checked(n)`

**Comment by andralex** [comment]
@WalterBright I couldn't edit this PR. Please rebase and mind the review - thx!

**Comment by JackStouffer** [comment]
As this is dead in the water, I've revived the changes here: https://github.com/dlang/phobos/pull/5990

**Pull request closed without merging**